File: toolbox/file-transfer/java/src/ft/Sender.java
/*
 * Copyright © Mihai Borobocea 2010
 * 
 * This file is part of File Transfer.
 * 
 * File Transfer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * File Transfer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with File Transfer.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */

package ft;

import static ft.Util.ascii;
import static ft.Util.readASCIILine;
import static ft.Util.transferBytes;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

class Sender {

	static void run(InputStream sockIn, OutputStream sockOut,
			List<String> fileArgs) throws IOException, BadASCIIException {
		List<FileInfo> fileInfos = getFileInfos(fileArgs);

		if (!announceFiles(sockIn, sockOut, fileInfos))
			return;

		sendFiles(sockIn, sockOut, fileInfos);
	}

	private static List<FileInfo> getFileInfos(List<String> fileArgs) {
		List<FileInfo> fileInfos = new ArrayList<FileInfo>(fileArgs.size());

		for (String fileArg : fileArgs) {
			File file = new File(fileArg);
			String name = file.getName();
			long size = file.length();

			fileInfos.add(new FileInfo(fileArg, name, size));
		}

		return fileInfos;
	}

	private static boolean announceFiles(InputStream sockIn,
			OutputStream sockOut, List<FileInfo> fileInfos) throws IOException,
			BadASCIIException {
		if (fileInfos.size() > 100) {
			System.err.println("Protocol can't transfer more than 100 files.");
			return false;
		}

		// start writing to socket

		String msg = "FILE COUNT " + fileInfos.size() + "\n";
		sockOut.write(msg.getBytes(ascii));

		for (int i = 0; i < fileInfos.size(); i++) {
			StringBuilder sb = new StringBuilder();
			sb.append(i + 1);
			sb.append("#\n");

			FileInfo crt = fileInfos.get(i);
			sb.append(crt.name);
			sb.append('\n');
			sb.append(crt.size);
			sb.append('\n');

			sockOut.write(sb.toString().getBytes(ascii));
		}

		// flush socket
		sockOut.flush();

		String expected = "SEND FILES\n";
		String got = readASCIILine(sockIn, expected.getBytes(ascii).length);
		if (!expected.equals(got)) {
			System.err.println("Receiver didn't show interest in files.");
			return false;
		}

		return true;
	}

	private static void sendFiles(InputStream sockIn, OutputStream sockOut,
			List<FileInfo> fileInfos) throws IOException, BadASCIIException {
		for (int i = 0; i < fileInfos.size(); i++) {
			String offerMsg = "OFFER " + (i + 1) + "\n";
			sockOut.write(offerMsg.getBytes(ascii));
			sockOut.flush();

			String yesMsg = "ACCEPT " + (i + 1) + "\n";
			String noMsg = "SKIP " + (i + 1) + "\n";
			int yesBytes = yesMsg.getBytes(ascii).length;
			int noBytes = noMsg.getBytes(ascii).length;
			String got = readASCIILine(sockIn, Math.max(yesBytes, noBytes));

			if (noMsg.equals(got)) {
				System.out.println(String.format("%-41.41s SKIPPED", fileInfos
						.get(i).name));
				continue;
			} else if (yesMsg.equals(got)) {
				boolean success = sendFileContents(sockOut, fileInfos.get(i));
				if (!success)
					return;

				String expected = "COMPLETED " + (i + 1) + "\n";
				got = readASCIILine(sockIn, expected.getBytes(ascii).length);
				if (!expected.equals(got)) {
					System.err.println("receiver didn't confirm completion");
					return;
				}
			} else {
				System.err.println("Invalid response to offer");
				return;
			}
		}
	}

	private static boolean sendFileContents(OutputStream sockOut,
			FileInfo fileInfo) throws IOException {
		InputStream fileIn = null;
		try {
			fileIn = new BufferedInputStream(new FileInputStream(
					fileInfo.cmdLineArg));

			long bytesSent = transferBytes(fileIn, sockOut, fileInfo.size,
					fileInfo.name);
			sockOut.flush();

			if (bytesSent != fileInfo.size) {
				System.err.println("Problem detected: file has " + bytesSent
						+ " bytes instead of " + fileInfo.size);
				return false;
			}

			return true;
		} finally {
			if (fileIn != null)
				fileIn.close();
		}
	}

}

File: toolbox/file-transfer/old-go/sender.go
package main

import (
	"os"
	"bufio"
	"strconv"
	"container/vector"
	"path"
	"strings"
	"fmt"
	"io"
)

func run_sender(sockIn *bufio.Reader, sockOut *bufio.Writer, fileArgs []string, fileSizes []int64) {
	var v vector.StringVector
	v.Push("FILE COUNT ")
	v.Push(strconv.Itoa(len(fileArgs)))
	v.Push("\n")

	for i := 0; i < len(fileArgs); i++ {
		v.Push(strconv.Itoa(i + 1))
		v.Push("#\n")
		v.Push(path.Base(fileArgs[i]))
		v.Push("\n")
		v.Push(strconv.Itoa64(fileSizes[i]))
		v.Push("\n")
	}
	announce := strings.Join(v, "")

	_, err := sockOut.Write([]byte(announce))
	if err == nil {
		err = sockOut.Flush()
	}
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return
	}

	expected := "SEND FILES"
	got, err := readAsciiLine(sockIn, len([]byte(expected))+1)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return
	}
	if got != expected {
		fmt.Println("Receiver didn't show interest in files")
		return
	}

	for i := 0; i < len(fileArgs); i++ {
		ok := sendFile(sockIn, sockOut, fileArgs, fileSizes, i)
		if !ok {
			return
		}
	}
}

func sendFile(sockIn *bufio.Reader, sockOut *bufio.Writer, fileArgs []string, fileSizes []int64, i int) bool {
	_, err := sockOut.Write([]byte("OFFER " + strconv.Itoa(i+1) + "\n"))
	if err == nil {
		err = sockOut.Flush()
	}
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return false
	}

	accMsg := "ACCEPT " + strconv.Itoa(i+1)
	skipMsg := "SKIP " + strconv.Itoa(i+1)
	var maxBytes int
	if len(accMsg) > len(skipMsg) {
		maxBytes = len([]byte(accMsg)) + 1
	} else {
		maxBytes = len([]byte(skipMsg)) + 1
	}

	got, err := readAsciiLine(sockIn, maxBytes)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return false
	}

	switch got {
	case accMsg:
		file, err := os.Open(fileArgs[i], os.O_RDONLY, 0)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return false
		}
		fileIn := bufio.NewReader(file)

		_, err = io.Copyn(sockOut, fileIn, fileSizes[i])
		if err == nil {
			err = sockOut.Flush()
		}
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return false
		}

		expected := "COMPLETED " + strconv.Itoa(i+1)
		got, err = readAsciiLine(sockIn, len([]byte(expected))+1)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return false
		}
		if got != expected {
			fmt.Println("Invalid response from receiver")
			return false
		}

		fmt.Println(fileArgs[i] + " DONE")
		return true
	case skipMsg:
		fmt.Println(fileArgs[i] + " SKIPPED")
		return true
	}

	fmt.Fprintf(os.Stderr, "Invalid reponse from receiver")
	return false
}
